kube/src/service/auth/oauth.rs

```diff
@@ -0,0 +1,106 @@
+use std::{env, path::PathBuf};
+
+#[cfg(feature = "rustls-tls")] use hyper_rustls::HttpsConnector;
+#[cfg(feature = "native-tls")] use hyper_tls::HttpsConnector;
+use tame_oauth::{
+    gcp::{ServiceAccountAccess, ServiceAccountInfo, TokenOrRequest},
+    Token,
+};
+
+use crate::{
+    error::{ConfigError, OAuthError},
+    Error, Result,
+};
+
+pub(crate) struct Gcp {
+    access: ServiceAccountAccess,
+    scopes: Vec<String>,
+}
+
+impl std::fmt::Debug for Gcp {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Gcp")
+            .field("access", &"{}".to_owned())
+            .field("scopes", &self.scopes)
+            .finish()
+    }
+}
+
+impl Gcp {
+    pub(crate) fn new(access: ServiceAccountAccess, scopes: Vec<String>) -> Self {
+        Self { access, scopes }
+    }
+
+    // Initialize ServiceAccountAccess so we can request later when needed.
+    pub(crate) fn from_env_and_scopes(scopes: Option<&String>) -> Result<Self> {
+        const DEFAULT_SCOPES: &str =
+            "https://www.googleapis.com/auth/cloud-platform,https://www.googleapis.com/auth/userinfo.email";
+        // Initialize ServiceAccountAccess so we can request later when needed.
+        let info = gcloud_account_info()?;
+        let access = ServiceAccountAccess::new(info).map_err(OAuthError::InvalidKeyFormat)?;
```

Comment:
`ServiceAccountAccess` has internal cache and returns `Request` when it's necessary to refresh the token. See `token` below.